closed/src/java.base/share/classes/openj9/internal/security/RestrictedSecurity.java

```diff
@@ -992,6 +982,40 @@ boolean isRestrictedProviderAllowed(String providerName) {
             return false;
         }
 
+        /**
+         * Get the provider name defined in provider construction method.
+         *
+         * @param providerName provider name or provider with packages or arguments
+         * @return provider name defined in provider construction method
+         */
+        private static String getProvidersSimpleName(String providerName) {
+            // Remove the provider's optional arguments if present.
+            int pos = providerName.indexOf(' ');
+            if (pos >= 0) {
+                providerName = providerName.substring(0, pos);
+            }
+            providerName = providerName.trim();
```

Comment:
I don't think this belongs here. Removing optional arguments should only occur in `initProviders()` so it does not alter the behavior of `isRestrictedProviderAllowed(String providerName)` in unwanted ways (if "SunPKCS11" is allowed, that doesn't mean "SunPKCS11 [some random text]" should be accepted there).

Reply:
Ok, I will move the below codes, back to the method `initProviders()`

```
            // Remove the provider's optional arguments if present.
            int pos = providerName.indexOf(' ');
            if (pos >= 0) {
                providerName = providerName.substring(0, pos);
            }
            providerName = providerName.trim();
```

And do you think the below trim for "-", will also need to move back to the `isRestrictedProviderAllowed()` ?

            // Remove argument, e.g. -NSS-FIPS, if present.
            pos = providerName.indexOf('-');
            if (pos >= 0) {
                providerName = providerName.substring(0, pos);
            }

Reply:
I think the removal of an argument (e.g. `-NSS-FIPS`) belongs in `initProviders()` as well, before the call to `getProvidersSimpleName()` instead of within `getProvidersSimpleName()`.

I don't think `isRestrictedServiceAllowed(Service service)` should be expecting arguments and so should not include code to remove them. However, it should be calling `getProvidersSimpleName()` so `SunSASL` is handled properly.

Reply:
> I think the removal of an argument (e.g. `-NSS-FIPS`) belongs in `initProviders()` as well, before the call to `getProvidersSimpleName()` instead of within `getProvidersSimpleName()`.
> 

The removal of an argument (e.g. -NSS-FIPS) can belongs in initProviders(). I will add the check before the call to `getProvidersSimpleName()` instead of within `getProvidersSimpleName()`.

> I don't think `isRestrictedServiceAllowed(Service service)` should be expecting arguments and so should not include code to remove them. However, it should be calling `getProvidersSimpleName()` so `SunSASL` is handled properly.

In the `isRestrictedServiceAllowed(Service service)`, it will get the provider name by the following code:

`String providerName = service.getProvider().getName();`

This `providerName` will always be the provider name defined in its construction method, for the `SunSASL` provider, it will alway return `SunSASL` as the `providerName`. So, I don't think it need to call the `getProvidersSimpleName()`.

Reply:
> In the `isRestrictedServiceAllowed(Service service)`, it will get the provider name by the following code:
> 
> `String providerName = service.getProvider().getName();`
> 
> This `providerName` will always be the provider name defined in its construction method, for the `SunSASL` provider, it will alway return `SunSASL` as the `providerName`. So, I don't think it need to call the `getProvidersSimpleName()`.

Agreed, but I think the code to remove arguments should be removed from `isRestrictedServiceAllowed(Service service)`.

Reply:
> > In the `isRestrictedServiceAllowed(Service service)`, it will get the provider name by the following code:
> > `String providerName = service.getProvider().getName();`
> > This `providerName` will always be the provider name defined in its construction method, for the `SunSASL` provider, it will alway return `SunSASL` as the `providerName`. So, I don't think it need to call the `getProvidersSimpleName()`.
> 
> Agreed, but I think the code to remove arguments should be removed from `isRestrictedServiceAllowed(Service service)`.

Here are the codes in isRestrictedServiceAllowed(Service service)

```
        boolean isRestrictedServiceAllowed(Service service) {
            String providerName = service.getProvider().getName();

            // Provider with argument, remove argument.
            // e.g. SunPKCS11-NSS-FIPS, remove argument -NSS-FIPS.
            int pos = providerName.indexOf('-');
            providerName = (pos < 0) ? providerName : providerName.substring(0, pos);

            Constraint[] constraints = providerConstraints.get(providerName);
            .....
```

I don't think the remove argument (e.g. remove -NSS-FIPS) can be removed from `isRestrictedServiceAllowed(Service service)`.

Because, for the provider like `SunPKCS11`, the provider name comes from two parts, one is the provider name defined in its construction method, which is `SunPKCS11`. Another part comes from its configuration file `${java.home}/conf/security/nss.fips.cfg`, which is` NSS-FIPS`, listed below.

```
name = NSS-FIPS
nssLibraryDirectory = /usr/lib64
nssSecmodDirectory = /etc/pki/nssdb
nssDbMode = readOnly
nssModule = fips
```

So, once the `SunPKCS11` is initialized using its configuration file, its provider name will be `SunPKSC11-NSS-FIPS`.

But when we generate the `providersSimpleName` list in method `initProviders()`. The provider name we get from the `java.security` file, which is 

`RestrictedSecurity.NSS.140-2.jce.provider.1 = SunPKCS11 ${java.home}/conf/security/nss.fips.cfg`. 

At that time, the provider is not initialized yet, and there is no way we can know the provider name will be `SunPKCS11-NSS-FIPS` at that time. So, we remove the configure file part ( which is `${java.home}/conf/security/nss.fips.cfg`) and use `SunPKCS11` as the `providersSimpleName` and store it into the `providersSimpleName` list.

But in the `isRestrictedServiceAllowed(Service service)`, the `providerName` returned in this method, is the provider which is already initialized, so the `providerName` will be `SunPKCS-NSS-FIPS`.  So, that is why we need the remove argument (e.g. remove -NSS-FIPS) here in this `isRestrictedServiceAllowed(Service service)`.

Reply:
I thought I saw a provider with a hyphen in its name, but when I looked again, I couldn't find it. So, yes, you're right `isRestrictedServiceAllowed(Service service)` should not change.

Reply:
Ok, then, I will move the below codes, out of `getProvidersSimpleName()`,  back to the method `initProviders()`

```
            // Remove the provider's optional arguments if present.
            int pos = providerName.indexOf(' ');
            if (pos >= 0) {
                providerName = providerName.substring(0, pos);
            }
            providerName = providerName.trim();

            // Remove argument, e.g. -NSS-FIPS, if present.
            pos = providerName.indexOf('-');
            if (pos >= 0) {
                providerName = providerName.substring(0, pos);
            }
```

And also, move the below code, back to the method `isRestrictedProviderAllowed()`.

            // Remove argument, e.g. -NSS-FIPS, if present.
            pos = providerName.indexOf('-');
            if (pos >= 0) {
                providerName = providerName.substring(0, pos);
            }

Reply:
@keithc-ca Codes updated according to above. Please help to review and advise.